codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java

```diff
@@ -95,28 +107,126 @@ public String toString() {
         return uriBuilder.toString();
     }
 
-    private static void appendComponent(String s, String charset, StringBuilder sb) {
-        try {
-            s = URLEncoder.encode(s, charset);
-        } catch (UnsupportedEncodingException ignored) {
-            throw new UnsupportedCharsetException(charset);
-        }
-        // replace all '+' with "%20"
-        int idx = s.indexOf('+');
-        if (idx == -1) {
-            sb.append(s);
-            return;
+    /**
+     * Encode the String as per RFC 3986, Section 2.
+     * <p>
+     * There is a little different between the JDK's encode method : {@link URLEncoder#encode(String, String)}.
+     * The JDK's encoder encode the space to {@code +} and this method directly encode the blank to {@code %20}
+     * beyond that , this method reuse the {@link #uriBuilder} in this class rather then create a new one,
+     * thus generates less garbage for the GC.
+     *
+     * @param s The String to encode
+     */
+    private void encodeNonUtf8Component(CharSequence s) {
+        //allocate memory until needed
+        char[] buf = null;
+
+        for (int i = 0; i < s.length();) {
+            char c = s.charAt(i);
+            if (dontNeedEncoding(c)) {
+                uriBuilder.append(c);
+                i++;
+            } else {
+                int index = 0;
+                if (buf == null) {
+                    buf = new char[s.length() - i];
+                }
+
+                do {
+                    buf[index] = c;
+                    index++;
+                    i++;
+                } while (i < s.length() && !dontNeedEncoding(c = s.charAt(i)));
+
+                byte[] bytes = new String(buf, 0, index).getBytes(charset);
+
+                for (byte b : bytes) {
+                    appendEncoded(b);
+                }
+            }
         }
-        sb.append(s, 0, idx).append("%20");
-        int size = s.length();
-        idx++;
-        for (; idx < size; idx++) {
-            char c = s.charAt(idx);
-            if (c != '+') {
-                sb.append(c);
+    }
+
+    /**
+     * @see ByteBufUtil#writeUtf8(io.netty.buffer.ByteBuf, CharSequence, int, int)
+     */
+    private void encodeUtf8Component(CharSequence s) {
```

Comment:
I would inline this directly into the caller and will pack the non-ascii parts as a slow path: it would use less inlining budget and will make the fast path (ascii?) inlined and faster...
I've made a similar optimization here https://github.com/netty/netty/pull/7539

Reply:
I will run some benchmark when I back home